File: word_chain.java
import java.util.*;

class WordChain {
	int N;
	Set<String> D;
	boolean[] alive;
	private String bef;
	private int P;

	public WordChain(int n, Set<String> d) {
		N = n;
		D = d;
		alive = new boolean[n];
		for (int i = 0; i < n; i++)
			alive[i] = true;
		bef = "";
		P = 0;
	}

	public void wordChain(String s) {
		if (D.contains(s) && ("".equals(bef) || bef.charAt(~-bef.length()) == s.charAt(0))
				&& s.charAt(~-s.length()) != 'z') {
			D.remove(s);
			bef = s;
		} else {
			alive[P] = false;
			bef = "";
		}
		do {
			P = -~P % N;
		} while (!alive[P]);
	}

	int aliveCount() {
		int a = 0;
		for (boolean b : alive)
			a += (b ? 1 : 0);
		return a;
	}

	boolean isAlive(int i) {
		return alive[i];
	}

}

public class Main {
	public static void main(String[] args) {
		Scanner sc = new Scanner(System.in);
		int N = sc.nextInt();
		int K = sc.nextInt();
		int M = sc.nextInt();
		Set<String> D = new HashSet<String>();
		while (K-- > 0)
			D.add(sc.next());
		WordChain wordChain = new WordChain(N, D);
		while (M-- > 0)
			wordChain.wordChain(sc.next());
		sc.close();
		System.out.println(wordChain.aliveCount());
		for (int i = 0; i < N; i++)
			if (wordChain.isAlive(i))
				System.out.println(-~i);

	}
}

File: word_chain.kt
import java.util.*

class WordChain(val n: Int, val d: MutableSet<String>) {
    private val alive = BooleanArray(n) { true }
    private var bef = ""
    private var p = 0

    fun wordChain(s: String) {
        if (d.contains(s) && (bef.isEmpty() || bef.last() == s.first()) && s.last() != 'z') {
            d.remove(s)
            bef = s
        } else {
            alive[p] = false
            bef = ""
        }
        do {
            p = (p + 1) % n
        } while (!alive[p])
    }

    fun aliveCount(): Int {
        return alive.count { it }
    }

    fun isAlive(i: Int): Boolean {
        return alive[i]
    }
}

fun main() {
    val sc = Scanner(System.`in`)
    val n = sc.nextInt()
    val k = sc.nextInt()
    val m = sc.nextInt()
    val d = mutableSetOf<String>()
    repeat(k) {
        d.add(sc.next())
    }
    val wordChain = WordChain(n, d)
    repeat(m) {
        wordChain.wordChain(sc.next())
    }
    sc.close()

    println(wordChain.aliveCount())
    for (i in 0 until n) {
        if (wordChain.isAlive(i)) {
            println(i + 1)
        }
    }
}
